pkg/proxy/proxy.go

```diff
@@ -0,0 +1,145 @@
+	// All paths allowed
+	if len(p.allowedPaths) == 0 {
+		return true
+	}
+
+	// Check if given passed exists in allowedPaths
+	for _, p := range p.allowedPaths {
+		if strings.TrimSuffix(strings.TrimSpace(p), "/") ==
+			strings.TrimSuffix(strings.TrimSpace(path), "/") {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *Proxy) redirect(hook *providers.Hook, path string) (gorequest.Response, []error) {
+	if hook == nil {
+		return nil, []error{errors.New("Cannot redirect with nil Hook")}
+	}
+	// Set SetDoNotClearSuperAgent to true so that request
+	// agent is not reset on POST call
+	request := gorequest.New().SetDoNotClearSuperAgent(true)
+
+	// Set Headers from hook
+	for key, value := range hook.Headers {
+		request.AppendHeader(key, value)
+	}
+
+	// Parse url to check validity
+	url, err := url.Parse(p.upstreamURL + path)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	// Assign default scheme as http if not specified
+	if url.Scheme == "" {
+		url.Scheme = "http"
+	}
+
+	resp, _, errs := request.Post(url.String()).Send(hook.Payload).End()
+
+	return resp, errs
+}
+
+func (p *Proxy) proxyRequest(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+
+	if !p.isPathAllowed(r.URL.Path) {
+		log.Printf("Not allowed to proxy path: '%s'", r.URL.Path)
+		http.Error(w, "Not allowed to proxy path: '"+r.URL.Path+"'", http.StatusForbidden)
+		return
+	}
+
+	provider, err := providers.NewProvider(p.provider, p.secret)
+	if err != nil {
+		log.Printf("Error creating provider: %s", err)
+		http.Error(w, "Error creating Provider", http.StatusInternalServerError)
+		return
+	}
+
+	hook, err := parser.Parse(r, provider)
+	if err != nil {
+		log.Printf("Eror Parsing Hook: %s", err)
+		http.Error(w, "Error parsing Hook: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if !provider.Validate(*hook) {
+		log.Printf("Eror Validating Hook: %s", err)
+		http.Error(w, "Error validating Hook", http.StatusBadRequest)
+		return
+	}
+
+	resp, errs := p.redirect(hook, r.URL.Path)
+	if errs != nil {
+		log.Printf("Error Redirecting '%s' to upstream '%s': %s\n", r.URL.Path, p.upstreamURL, errs)
+		http.Error(w, "Error Redirecting '"+r.URL.Path+"' to upstream '"+p.upstreamURL+"'", http.StatusInternalServerError)
+		return
+	}
+
+	if resp.StatusCode >= 400 {
+		log.Printf("Error Redirecting '%s' to upstream '%s', Upstream Redirect Status: %s\n", r.URL.Path, p.upstreamURL, resp.Status)
+		http.Error(w, "Error Redirecting '"+r.URL.Path+"' to upstream '"+p.upstreamURL+"' Upstream Redirect Status:"+resp.Status, resp.StatusCode)
+		return
+	}
+
+	log.Printf("Redirected incomming request '%s' to '%s' with Response: '%s'\n",
+		r.URL.Path, p.upstreamURL, resp.Status)
+}
+
+// Health Check Endpoint
+func (p *Proxy) health(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+	w.WriteHeader(200)
+	w.Write([]byte("I'm Healthy and I know it! ;) "))
+}
+
+func NewProxy(listenAddress string, upstreamURL string, allowedPaths []string,
```

Comment:
I think we should just create and return a new proxy object. Proxy will then have Run() method which will create the router and serve, basically the last 5 lines of this method. The reason why it should be like this is because methods starting with New are supposed to be constructors. Also writing NewProxy in main function doesnt make sense to me.
We can do like this in main:
```
proxy := NewProxy(params)

if err := proxy.Run(); err != nil {
  Log.fatal(err)
}
```

Run method will have the following last line `return http.ListenAndServe(args)`

Reply:
Correct, will do